Pass on iframe scripts (#761)
browserengineering/book: book/embeds.md

Iframe scripts
==============

We've now got users interacting with iframes directly. The last step
in adding iframe support, then, is allowing some kind of programmatic
access. Of course, each frame can _already_ run scripts---but right
now, each `Frame` has its own `JSContext`, so these scripts can't
really interact with each other. In reality, *same-origin* iframes run
in the same JavaScript context and can access each other's globals,
call each other's functions, and modify each other's DOMs. Let's
implement that.

For two frames' JavaScript environments to interact, we'll need to put
them in the same `JSContext`. So, instead of each `Frame` having a

> `JSContext` of its own, we'll want to store `JSContext`s on the `Tab`:

Reviewer: ```suggestion
`JSContext` of its own, we'll want to store `JSContext`s on the `Tab`, in a
dictionary that maps from origin to JS context:
```

``` {.python}
class Tab:
    def __init__(self, browser):
        # ...
        self.origin_to_js = {}

    def get_js(self, origin):
        if origin not in self.origin_to_js:
            self.origin_to_js[origin] = JSContext(self)
        return self.origin_to_js[origin]
```

Each `Frame` will then ask the `Tab` for its JavaScript context:

``` {.python}
class Frame:
    def load(self, url, body=None):
        # ...
        self.js = self.tab.get_js(url_origin(url))
        # ...
```

> We've now got multiple page's scripts living inside one JavaScript

Reviewer: ```suggestion
We've now got multiple pages' scripts living inside one JavaScript
```

refers to the [global object][global-object], and instead of writing a
global variable like `a`, you can always write `window.a` instead. To
keep our implementation simple, in our browser, scripts will always
need to reference variable and functions via `window`. We'll need to
do the same in our runtime:

``` {.js}
window.console = { log: function(x) { call_python("log", x); } }

// ...

window.Node = function(handle) { this.handle = handle; }

// ...
```

Do the same for every function or variable in the `runtime.js` file.
If you miss one, you'll get errors like this:

    _dukpy.JSRuntimeError: ReferenceError: identifier 'Node' undefined
    	duk_js_var.c:1258
    	eval src/pyduktape.c:1 preventsyield

Then you'll need to go find where you forgot to put `window.` in front
of `Node`.

::: {.quirk}
Demos from previous chapters will need to be similarly fixed up before
they work. For example, `setTimeout` might need to change to
`window.setTimeout`, etc.
:::

[global-object]: https://developer.mozilla.org/en-US/docs/Glossary/Global_object

To get multiple frames' scripts to play nice inside one JavaScript
context, we'll create multiple `Window` objects, so imagine having a

> `window1`, a `window2`, and so on. Before running a frame's scripts,

Reviewer: ```suggestion
`window_1`, a `window_2`, and so on. Before running a frame's scripts,
```
